### ISSUE TITLE
Simplify the tests fixture system

### ISSUE THREAD

**Issue by LefterisJP** [raiden-network/raiden#3347]
## Problem Definition

[Suggested](https://github.com/raiden-network/raiden/tree/test-review/ob-review#simplifying-the-fixture-system) to us during our test review. 

Our current fixture system uses fixtures even for constant values that can be parameterized. This leads to a very complicated and deep fixture interdependency tree. The end result is that all fixtures are stuck with the lowest common denominator fixture scope which is the `function` scope.

## Motivation for refactoring

- Configuration of a test does not force all fixtures used by that test into having function scope anymore
- Making explicit what is going on, makes the tests easier to understand

## Suggestion

Follow these steps:

### Step 1

Identify fixtures that are really just constants and are never overwritten and turn them into constants. This might just be a handful of fixtures, but it is a handful of fixtures less in an already quite impressive dependency tree.

### Step 2

Collect all fixtures that are actually per-test-configuration (mainly for one of the "god fixtures" like `raiden_network`) into a configuration object that has sensible defaults (which are now the original return value of the fixtures). E.g. `number_of_nodes` and `channels_per_node`.

This could also take care of the `skip_*` fixtures that are used to filter out tests already on collection time.

### Step 3

Use a different way to pass this config object to the fixture that needs it. The obvious choice here is to make more extensive use of `pytest_generate_tests` which is already used like this to skip tests and to parametrize `transport` and `private_rooms`.

Rather than having a ton of fixtures all taking care of one aspect of the test configuration there would be one configuration object taking care of all these aspects (should it be skipped under certain circumstances? Which kind of dynamic parametrization is needed, which concrete settings need to be passed to the fixture(s) using them, etc.)

This should also be well documented as this is a non obvious use of the testing mechanisms to people who are not deeply into how pytest and the tests themselves work.

**Comment by hackaugusto** [comment]
I agree that we have a huge number of fixtures and that understanding it at first may be daunting. However, using a configuration object will just move the "dautinguiness" around, from the pytest fixtures to a configuration object with lots of variables. One advantage may be to have all the configurations in a single configuration object, however, I would argue that this is what the [fixtures/variables.py](https://github.com/raiden-network/raiden/blob/a5e326b81635e2aac1967a645cd8b8283ba5e1de/raiden/tests/fixtures/variables.py) module is for.

With that said, I'm not against this change, but I like it for a different reason. I would like to have test code that does not depend on pytest to run. The problem we have right now with the slow integration tests is because we run these sequentially, most of the time these tests are just waiting.

This happened for a few reasons, the biggest one is that the integration tests are written with a clean slate mindset, meaning most tests expect to start with a fresh token network smart contract, so that it can know the world is of its own and do things like count the number of open channels. Because we deploy smart contract with [smart contract fixtures](https://github.com/raiden-network/raiden/blob/a5e326b81635e2aac1967a645cd8b8283ba5e1de/raiden/tests/integration/fixtures/smartcontracts.py) we are tied to the execution model from pytest, which is one test at the time. The fix here would be to have a description object for each test, with the environment the test expect to run with, once we have these descriptions we can setup and start multiple tests in parallel.

For the above I tried an experiment a while back, the idea was to have a [spec for each test](https://github.com/hackaugusto/raiden/commit/f54bc1bfdec5f831e0f713ac9e6a780803b0916d#diff-1d1f900c3b149a25fc8cf4650978ff8bR45), each spec describe the environment for one test and are used [to deploy the smart contract](https://github.com/hackaugusto/raiden/commit/f54bc1bfdec5f831e0f713ac9e6a780803b0916d#diff-1d1f900c3b149a25fc8cf4650978ff8bR207) (would be done in parallel), after this is done then [each test is executed in parallel](https://github.com/hackaugusto/raiden/commit/f54bc1bfdec5f831e0f713ac9e6a780803b0916d#diff-1d1f900c3b149a25fc8cf4650978ff8bR472).

The prototype probably does not work anymore, and can be improved, but that is my idea. This would probably make the CI way faster, but potentially harder to debug / flakier, for this I would suggest a two step run, first run everything in parallel and collect failures, for each failure run the test again in isolation, if it fails again then fail the build.

**Comment by LefterisJP** [comment]
I am torn about this.

I don't see the benefit of implementing the original issue suggestion when compared with the amount of work it would take.

Doing what Augusto suggested above at least has a clear goal: **make integration test runs faster**. We don't need integration tests to have a clean slate each time. The scenario player already does that so we could go with the suggestion.

I would consider exploring this but it is going to be a considerable amount of work so it needs discussion.
